### PR TITLE
fast float/double writer option

### PR DESCRIPTION
**Pull request by pjfanning**
relates to https://github.com/FasterXML/jackson-core/issues/514

### CODE REVIEW

File: src/test/java/com/fasterxml/jackson/core/io/schubfach/FloatToStringTest.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    assertF2sEquals("4103.9004", 4103.9003f);
+    assertF2sEquals("5.3399997E9", 5.3399997E9f);
+    assertF2sEquals("6.0898E-39", 6.0898E-39f);
+    assertF2sEquals("0.0010310042", 0.0010310042f);
+    assertF2sEquals("2.882326E17", 2.8823261E17f);
+    assertF2sEquals("7.038531E-26", 7.038531E-26f);
+    assertF2sEquals("9.223404E17", 9.2234038E17f);
+    assertF2sEquals("6.710887E7", 6.7108872E7f);
+    //TODO investigate
+    assertF2sEquals("9.8E-45", 1.0E-44f);
```
Comment (pjfanning) [inline src/test/java/com/fasterxml/jackson/core/io/schubfach/FloatToStringTest.java:92]: @plokhotnyuk I took some of the Ryu tests and used them to test Schubfach and this line 92 is one strange diff - the assert works as "1.0E-44f" in Ryu but I had to hack the value here for Schubfach. Any ideas?

Comment (plokhotnyuk) [reply src/test/java/com/fasterxml/jackson/core/io/schubfach/FloatToStringTest.java:92]: Results from Schubfach are rounded with exact precision of 2 decimal digits.

Just see exact representation of float with significant 7:
https://float.exposed/0x00000007

**Comment by pjfanning** [comment]
@cowtowncoder the Ryu and Schubfach algorithms don't output exactly the same values as JDK Double.toString - at least for some edge cases - would there be anyway to have jackson-core NumberOutput support JDK approach as default and to have an option to use Schubfach (which is faster than Ryu) if the user enables a flag?

**Comment by cowtowncoder** [comment]
@pjfanning I think it is reasonable to have a non-default flag to enable "fast but incorrect" (I think? JDK implementations is, from what I recall, quite precisely compatible well-defined criteria of closest value, round-trupping). As with `NumberInput` there is the question of how to handle configurability; I'm pretty sure it should default to JDK for legacy cases where boolean setting cannot be passed.

**Comment by pjfanning** [comment]
@cowtowncoder I'll probably hold back on doing anything with this writer code until the fast double parser work is merged. I think that is getting close enough to something that could be considered for merging.

**Comment by cowtowncoder** [comment]
Ok I think this can be closed?

**Comment by pjfanning** [comment]
@cowtowncoder the recent merges affect parsing (and deserialization) - this PR affects serialization - there is still a little bit of work to do on this but it is close enough to being ready

File: src/main/java/com/fasterxml/jackson/core/json/JsonWriteFeature.java [hunk 2]
```diff
@@ -72,6 +72,15 @@
...
     ESCAPE_NON_ASCII(false, JsonGenerator.Feature.ESCAPE_NON_ASCII),
 
+    /**
+     * Feature that determines whether to use standard Java code to write floats/doubles (default) or
+     * use the Schubfach algorithm which is faster. The latter approach may lead to small
+     * differences in the precision of the float/double that is written to the JSON output.
+     *
+     * @since 2.14
+     */
+    USE_FAST_DOUBLE_WRITER(false, JsonGenerator.Feature.USE_FAST_DOUBLE_WRITER),
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/json/JsonWriteFeature.java:82]: As with reader side, let's move this to more general `StreamWriteFeature` as it need not be JSON specific but applicable to most if not all textual format backends.

**Comment by cowtowncoder** [comment]
@pjfanning d'oh! I really should read the description in detail before commenting something stupid. Yes, that makes sense.
Just let me know when merging would make sense. Could be split in multiple parts too but may not be necessary.

One thing that makes things easier to merge is that 3.0 (master) has immutable factories, so using builders (JsonFactory.builder().... or equivalent from `BaseTest`) avoids some merging issues. But I can usually handle those quite easily, just cause intermediate merge conflict.

**Comment by pjfanning** [comment]
@cowtowncoder is this PR now a manageable size? It will have some merge issues in master but shouldn't be too bad.

**Pull request merged**